### PR TITLE
[WIP]Metric registry

### PR DESCRIPTION
**Pull request by hasetz**
MetricRegistry works in shell, but not with purest --> Need to investigate further.

### CODE REVIEW

**Comment by MBrouns** [comment]
You mean pytest instead of purest?

**Comment by hasetz** [comment]
> You mean pytest instead of purest?

Jup :)

File: src/skprometheus/metrics.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+from skprometheus.prom_client_utils import add_labels
+
+
+class _MetricRegistry:
+    """Object for initiation and upkeep of metrics. Necessary for avoiding assignment and label conflicts."""
+    def __init__(self):
+        self.labels = set()
+        self.metrics_initialized = False
+        self.current_labels = {}
+        self.DEFAULT_LATENCY_BUCKETS = (0.005, 0.01, 0.025, 0.05, 0.075, 0.1, 0.25,
```
Comment (MBrouns) [inline src/skprometheus/metrics.py:11]: We'll probably want a way for users to set these

File: src/skprometheus/metrics.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+                                        0.5, 0.75, 1., 2.5, 5., 7.5, 10., float('inf'))
+        self.DEFAULT_PROBA_BUCKETS = (0, 0.1, 0.2, 0.3, 0.4, 0.5, 0.6, 0.7, 0.8, 0.9, 1)
+
+    def set_labels(self, labels):
+        if self.metrics_initialized:
```
Comment (MBrouns) [inline src/skprometheus/metrics.py:16]: I'd like a test to see if this exception gets thrown correctly

File: src/skprometheus/metrics.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        self.current_labels = labels
+        return self
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, type, value, traceback):
+        self.current_labels = {}
+
+    def model_predict_total(self):
```
Comment (MBrouns) [inline src/skprometheus/metrics.py:58]: These functions will have a lot of repetition, especially when we add more and more metrics. We might want to get a bit smart with implementing a `__getattr__` at some point

File: src/skprometheus/metrics.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+        if self.metrics_initialized:
+            raise ValueError('can only add labels before metrics are initialized')
+
+        self.labels = set.union(self.labels, labels)
+
+    def _init_metrics(self):
+        if self.metrics_initialized:
+            return
+        self.metrics_initialized = True
+        self._model_predict = Counter(
```
Comment (MBrouns) [inline src/skprometheus/metrics.py:25]: I'm not sure how I feel about defining **all** the metrics we'll ever need in here. We might want to allow some kind of external registration of metrics

**Comment by MBrouns** [comment]
I added some comments but those don't need to be addressed right now, they're more of a reminder for the future

**Pull request merged**